<center>

$\Huge \textbf{Universidad Nacional Autónoma de México}$  
$\Huge \textbf{Facultad de Ciencias}$  
<p align="center">
  <img src="https://www.icat.unam.mx/wp-content/uploads/2021/11/Copia-de-LogoUNAM.-Azul.-Fondo-transparente.png" alt="UNAM" width="200"/>
</p>

<hr style="height:3px; background-color:#0B6E4F; border:none;"/>


$\LARGE \textbf{Inteligencia Artificial}$  

$\Large \textit{Laboratorio 5.8}$  


\begin{array}{rl}
\textbf{Docente:} & Dra. Jessica Sarahi Méndez Rincón \\[6pt]
\textbf{Ayudante de laboratorio:} & Diego Eduardo Peña Villegas \\[6pt]
\textbf{Alumno:} & Isaac Giovani Escobar Gonzalez \\[6pt]
\textbf{Fecha de realización:} & 06/05/2026
\end{array}

</center>


**UNAM y la creadora del presente notebook, no guarda relación alguna con las marcas mencionadas como ejemplo. Las marcas son propiedad de sus titulares conforme a la legislación aplicable, se utilizan con fines académicos y didácticos, por lo que no existen fines de lucro, relación publicitaria o de patrocinio.

---

In [1]:
!pip install transformers

In [2]:
from transformers import pipeline, AutoTokenizer, AutoModelForSequenceClassification

# Cargar el modelo pre-entrenado y el tokenizador de BERT en español
#model_name = "dccuchile/bert-base-spanish-wwm-uncased"
# Este ya sabe qué es Positivo, Negativo y Neutral
# Muy robusto para español
model_name = "nlptown/bert-base-multilingual-uncased-sentiment"
tokenizer = AutoTokenizer.from_pretrained(model_name)
model = AutoModelForSequenceClassification.from_pretrained(model_name)

# Crear un pipeline para la clasificación de texto
text_classification_pipeline = pipeline(
    'text-classification',
    model=model,
    tokenizer=tokenizer
)

# Ejecutar la clasificación de texto en una oración en español
#result = text_classification_pipeline('Esta película es muy buena, la recomiendo totalmente')
#print(result)

# Ejecutar la clasificación
result = text_classification_pipeline('Esta película es muy buena, la recomiendo totalmente')

# Acceder a los datos dentro de la lista y el diccionario
label = result[0]['label']
score = result[0]['score']

print(f"Resultado: {label}")
print(f"Confianza: {score:.4f}")

/usr/local/lib/python3.12/dist-packages/huggingface_hub/utils/_auth.py:93: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


config.json:   0%|          | 0.00/953 [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/39.0 [00:00<?, ?B/s]

vocab.txt: 0.00B [00:00, ?B/s]

special_tokens_map.json:   0%|          | 0.00/112 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/669M [00:00<?, ?B/s]

Loading weights:   0%|          | 0/201 [00:00<?, ?it/s]

Resultado: 5 stars
Confianza: 0.7252


1. El cambio en las etiquetas (Labels)
Al usar un modelo ajustado (probablemente el de nlptown o uno similar de reseñas), el sistema ya no usa códigos genéricos como LABEL_1.

"5 stars": El modelo ha clasificado la frase ("Esta película es muy buena, la recomiendo totalmente") con la puntuación máxima. Ha detectado un sentimiento altamente positivo.

2. La Confianza (Score)
0.7252 (72.5%): Es una métrica sólida. Indica que, de todas las opciones posibles (1, 2, 3, 4 o 5 estrellas), el modelo está bastante seguro de que pertenece a la categoría de 5.

In [3]:
from transformers import AutoTokenizer, AutoModelForCausalLM

# Descargar el modelo pre-entrenado de GPT-2 en español
model_name = "mrm8488/spanish-gpt2"
tokenizer = AutoTokenizer.from_pretrained(model_name)
model = AutoModelForCausalLM.from_pretrained(model_name)

config.json:   0%|          | 0.00/883 [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/226 [00:00<?, ?B/s]

vocab.json: 0.00B [00:00, ?B/s]

merges.txt: 0.00B [00:00, ?B/s]

added_tokens.json:   0%|          | 0.00/24.0 [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/90.0 [00:00<?, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

model.safetensors:   0%|          | 0.00/510M [00:00<?, ?B/s]

Loading weights:   0%|          | 0/148 [00:00<?, ?it/s]

GPT2LMHeadModel LOAD REPORT from: mrm8488/spanish-gpt2
Key                                     | Status     |  | 
----------------------------------------+------------+--+-
transformer.h.{0...11}.attn.masked_bias | UNEXPECTED |  | 
transformer.h.{0...11}.attn.bias        | UNEXPECTED |  | 

Notes:
- UNEXPECTED	:can be ignored when loading from different task/architecture; not ok if you expect identical arch.


In [5]:
#Cambiar este texto para 3 ejemplos
input_str = "Había una vez"

max_length = 100
# Generar texto con el modelo pre-entrenado de GPT-2 en español
input_ids = tokenizer.encode(input_str, return_tensors="pt")
sample_outputs = model.generate(
    input_ids,
    do_sample=True,
    max_length=max_length,
    top_k=50,
    top_p=0.95,
    temperature=1.0,
    num_return_sequences=1
)
generated_text = tokenizer.decode(sample_outputs[0], skip_special_tokens=True)



print(generated_text)

The attention mask and the pad token id were not set. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.
Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.


Había una vez un hombre con problemas, que se creía un dios.Ahora es un borracho y está enojado y ha cambiado de idea.Hay alguien con quien podría hablar.No hay nadie.Si quiere hablar con alguien, me gustaría estar a solas con ella.¿Y bien?¿Qué?Me ha dejado.¿Cómo lo sabes?Lo he visto en mis sueños.-¿Dónde?-En la calle.No está aquí.Lo vi cuando volví.Entonces sabes que no está aquí.Está


In [6]:
#Cambiar este texto para 3 ejemplos
input_str = "El hombre bicentenario es"

max_length = 100
# Generar texto con el modelo pre-entrenado de GPT-2 en español
input_ids = tokenizer.encode(input_str, return_tensors="pt")
sample_outputs = model.generate(
    input_ids,
    do_sample=True,
    max_length=max_length,
    top_k=50,
    top_p=0.95,
    temperature=1.0,
    num_return_sequences=1
)
generated_text = tokenizer.decode(sample_outputs[0], skip_special_tokens=True)



print(generated_text)

The attention mask and the pad token id were not set. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.
Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.


El hombre bicentenario es su nuevo reto.Cuando quiera ayuda no encontrará a otro supervisor del que no me haya hablado.Buenas noches, Sr. Garrison.Traducción:Elena Barcia con "Memory S23".En breve, un hombre con una pistola desconocida va a entrar a mi espacio.Tengo una pregunta.En el mundo real, ¿quién es este hombre?Este monstruo, yo creo que se llama, ¿me entiendes?Digo, él puede hacer lo que sea, ¿no?


In [7]:
#Cambiar este texto para 3 ejemplos
input_str = "El volcan de Parangaricutirimicuaro"

max_length = 100
# Generar texto con el modelo pre-entrenado de GPT-2 en español
input_ids = tokenizer.encode(input_str, return_tensors="pt")
sample_outputs = model.generate(
    input_ids,
    do_sample=True,
    max_length=max_length,
    top_k=50,
    top_p=0.95,
    temperature=1.0,
    num_return_sequences=1
)
generated_text = tokenizer.decode(sample_outputs[0], skip_special_tokens=True)



print(generated_text)

The attention mask and the pad token id were not set. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.
Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.


El volcan de Parangaricutirimicuaro, al este del estado Amazonas, es la caldera del gran lago Purus, en el sudeste de Brasil, el sitio en el cual se encuentran varios de los yacimientos de petróleo más grandes del mundo.A 1.5 km al sur del lago Purus, las cimas de la cadena de volcanes del norte de Brasil, los volcanes de la Serra de los Açus y de la Serra de la Mata, el cerro de


# Conclusión

En esta práctica se evaluaron dos arquitecturas transformer aplicadas al procesamiento del español: un modelo BERT multilingüe ajustado para análisis de sentimiento y un modelo GPT-2 entrenado en español. El modelo BERT clasificó correctamente una expresión positiva con una etiqueta numérica (5 estrellas) y una confianza del 72.5%, demostrando su capacidad para tareas de clasificación supervisada con granularidad en la intensidad del sentimiento.

Por su parte, el modelo GPT-2 generó continuaciones textuales coherentes a partir de semillas cortas, aunque con limitaciones típicas de modelos autorregresivos no finamente ajustados, como cierta pérdida de coherencia temática. En conjunto, los experimentos confirman que ambos modelos son funcionales en español y permiten diferenciar claramente entre tareas de comprensión (BERT) y generación (GPT-2) dentro del ecosistema de Transformers.

<hr/>
<footer style="text-align:center; font-size:12px; color:gray;">
© 2026 UNAM Facultado de Ciencias – Todos los derechos reservados

</footer>